src/main/java/appeng/fluids/util/AEFluidInventory.java

```diff
@@ -0,0 +1,377 @@
+	{
+		final IAEFluidStack fluid = this.fluids[slot];
+		if( fluid == null || maxDrain <= 0 )
+		{
+			return null;
+		}
+
+		int drained = maxDrain;
+		if( fluid.getStackSize() < drained )
+		{
+			drained = (int) fluid.getStackSize();
+		}
+
+		FluidStack stack = new FluidStack( fluid.getFluid(), drained );
+		if( doDrain )
+		{
+			fluid.setStackSize( fluid.getStackSize() - drained );
+			if( fluid.getStackSize() <= 0 )
+			{
+				this.fluids[slot] = null;
+			}
+			onContentChanged( slot );
+		}
+		return stack;
+	}
+
+	@Override
+	public int fill( final FluidStack fluid, final boolean doFill )
+	{
+		if( fluid == null || fluid.amount <= 0 )
+			return 0;
+
+		final FluidStack insert = fluid.copy();
+
+		int totalFillAmount = 0;
+		for( int slot = 0; slot < this.getSlots(); ++slot )
+		{
+			int fillAmount = this.fill( slot, insert, doFill );
+			totalFillAmount += fillAmount;
+			insert.amount -= fillAmount;
+			if( insert.amount <= 0 )
+				break;
+		}
+		return totalFillAmount;
+	}
+
+	@Override
+	public FluidStack drain( final FluidStack fluid, final boolean doDrain )
+	{
+		if( fluid == null || fluid.amount <= 0 )
+			return null;
+
+		final FluidStack resource = fluid.copy();
+
+		FluidStack totalDrained = null;
+		for( int slot = 0; slot < this.getSlots(); ++slot )
+		{
+			FluidStack drain = this.drain( slot, resource, doDrain );
+			if( drain != null )
+			{
+				if( totalDrained == null )
+					totalDrained = drain;
+				else
+					totalDrained.amount += drain.amount;
+
+				resource.amount -= drain.amount;
+				if( resource.amount <= 0 )
+					break;
+			}
+		}
+		return totalDrained;
+	}
+
+	@Override
+	public FluidStack drain( final int maxDrain, final boolean doDrain )
+	{
+		if( maxDrain == 0 )
+			return null;
+
+		FluidStack totalDrained = null;
+		int toDrain = maxDrain;
+
+		for( int slot = 0; slot < this.getSlots(); ++slot )
+		{
+			if( totalDrained == null )
+			{
+				totalDrained = this.drain( slot, toDrain, doDrain );
+				if( totalDrained != null )
+				{
+					toDrain -= totalDrained.amount;
+				}
+			}
+			else
+			{
+				FluidStack copy = totalDrained.copy();
+				copy.amount = toDrain;
+				FluidStack drain = this.drain( slot, copy, doDrain );
+				if( drain != null )
+				{
+					totalDrained.amount += drain.amount;
+					toDrain -= drain.amount;
+				}
+			}
+
+			if( toDrain <= 0 )
+				break;
+		}
+		return totalDrained;
+	}
+
+	public void writeToNBT( final NBTTagCompound data, final String name )
+	{
+		final NBTTagCompound c = new NBTTagCompound();
+		this.writeToNBT( c );
+		data.setTag( name, c );
+	}
+
+	private void writeToNBT( final NBTTagCompound target )
+	{
+		for( int x = 0; x < this.fluids.length; x++ )
+		{
+			try
+			{
+				final NBTTagCompound c = new NBTTagCompound();
+
+				if( this.fluids[x] != null )
+				{
+					this.fluids[x].writeToNBT( c );
+				}
+
+				target.setTag( "#" + x, c );
+			}
+			catch( final Exception ignored )
+			{
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Either remove or fill this block of code. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.orod.org/coding_rules#rule_key=squid%3AS00108)